src/bots/fillerBulk.ts

```diff
@@ -0,0 +1,180 @@
+
+		const fillSlot = this.orderSubscriber.getSlot();
+
+		return {
+			nodesToFill: this.findNodesToFill(market, dlob),
+			nodesToTrigger: dlob.findNodesToTrigger(
+				marketIndex,
+				fillSlot,
+				oraclePriceData.price,
+				MarketType.PERP,
+				this.driftClient.getStateAccount()
+			),
+		};
+	}
+
+	protected findNodesToFill(
+		market: PerpMarketAccount,
+		dlob: DLOB
+	): NodeToFill[] {
+		const nodesToFill = [];
+		const marketIndex = market.marketIndex;
+
+		const oraclePriceData =
+			this.driftClient.getOracleDataForPerpMarket(marketIndex);
+
+		const vAsk = calculateAskPrice(market, oraclePriceData);
+		const vBid = calculateBidPrice(market, oraclePriceData);
+
+		const updatedAMM = calculateUpdatedAMM(market.amm, oraclePriceData);
+		const vammCanFillLongs = calculateMaxBaseAssetAmountFillable(
+			updatedAMM,
+			PositionDirection.LONG
+		).gte(market.amm.minOrderSize);
+		const vammCanFillShorts = calculateMaxBaseAssetAmountFillable(
+			updatedAMM,
+			PositionDirection.SHORT
+		).gte(market.amm.minOrderSize);
+
+		const fillSlot = this.orderSubscriber.getSlot();
+
+		let bestBid = dlob.getBestBid(
+			marketIndex,
+			fillSlot,
+			MarketType.PERP,
+			oraclePriceData
+		);
+		// only consider vamm if it has liquidity
+		if (vammCanFillShorts) {
+			bestBid = bestBid ? BN.max(bestBid, vBid) : vBid;
+		}
+
+		const seenBidMaker = new Set<string>();
+		const restingBids = dlob.getRestingLimitBids(
+			marketIndex,
+			fillSlot,
+			MarketType.PERP,
+			oraclePriceData
+		);
+		const topRestingBids = [];
+		for (const restingBid of restingBids) {
+			topRestingBids.push(restingBid);
+			seenBidMaker.add(restingBid.userAccount?.toString() || '');
+			if (seenBidMaker.size == MAX_NUM_MAKERS) {
+				break;
+			}
+		}
+
+		let bestAsk = dlob.getBestAsk(
+			marketIndex,
+			fillSlot,
+			MarketType.PERP,
+			oraclePriceData
+		);
+		// only consider vamm if it has liquidity
+		if (vammCanFillLongs) {
+			bestAsk = bestAsk ? BN.min(bestAsk, vAsk) : vAsk;
+		}
+
+		const seenAskMaker = new Set<string>();
+		const restingAsks = dlob.getRestingLimitAsks(
+			marketIndex,
+			fillSlot,
+			MarketType.PERP,
+			oraclePriceData
+		);
+		const topRestingAsks = [];
+		for (const restingAsk of restingAsks) {
+			topRestingAsks.push(restingAsk);
+			seenAskMaker.add(restingAsk.userAccount?.toString() || '');
+			if (seenAskMaker.size == MAX_NUM_MAKERS) {
+				break;
+			}
+		}
+
+		if (bestAsk) {
+			const takingBidsGenerator = dlob.getTakingBids(
+				marketIndex,
+				MarketType.PERP,
+				fillSlot,
+				oraclePriceData
+			);
+
+			for (const takingBid of takingBidsGenerator) {
+				const takingBidPrice = takingBid.getPrice(oraclePriceData, fillSlot);
+				if (!takingBidPrice || takingBidPrice.gte(bestAsk)) {
+					nodesToFill.push({
+						node: takingBid,
+						makerNodes: topRestingAsks.filter(
+							(node) => !node.userAccount!.equals(takingBid.userAccount!)
+						),
+					});
+				}
+			}
+		}
```

Comment:
might be insightful to add an `else` here and log whenever there is no bestBid or bestAsk, the user, market index, etc.